repos/system_upgrade/el9toel10/actors/opensslenginescheck/libraries/opensslenginescheck.py

```diff
@@ -0,0 +1,154 @@
+
+        TLS.MaxProtocol = TLSv1.3
+        DTLS.MaxProtocol = DTLSv1.2
+    """
+    if pair.key == key:
+        return True
+    return _normalize_key(pair.key) == key
+
+
+def _find_pair(block, key):
+    """
+    Find key-value pair in the given configuration block
+
+    In the given configuration block (OpenSslConfigBlock) find a key-value with a given key.
+    If multiple values match, only the last one is returned.
+    """
+    res = None
+    for pair in block.pairs:
+        if _key_equal(pair, key):
+            res = pair
+
+    return res
+
+
+def _openssl_find_block(config, name):
+    """
+    In the given configuration file (OpenSslConfig) find a block with a given name
+    """
+    for block in config.blocks:
+        if block.name == name:
+            return block
+
+    return None
+
+
+resources = [
+    reporting.RelatedResource('package', 'openssl'),
+    reporting.RelatedResource('file', '/etc/pki/tls/openssl.cnf')
+]
+
+
+def check_openssl_engines(config):
+    """
+    Check there are no engines configured in openssl.cnf
+
+    In case the of the pkcs11 engine, stop the upgrade as the package was removed from RHEL 10.
+
+    In case of other (usually third-party) engines, just warn about them being deprecated,
+    and suggest removal.
+    """
+    init_block = _openssl_find_block(config, "openssl_init")
+    if config.openssl_conf != "openssl_init" or not init_block:
+        reporting.create_report([
+            reporting.Title('Non-standard configuration of openssl.cnf'),
+            reporting.Summary(
+                'The OpenSSL configuration file `/etc/pki/tls/openssl.cnf` does not contain '
+                'expected initialization (openssl_conf = openssl_init key-value pair).'
+            ),
+            reporting.Remediation(
+                'The openssl.cnf file needs to contain the following initialization: '
+                '`openssl_conf = openssl_init` The `openssl_conf` now contains {} or '
+                'the `[ openssl_init ]` block is missing. '.format(config.openssl_conf)
+            ),
+            reporting.Groups([reporting.Groups.INHIBITOR]),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([
+                    reporting.Groups.SECURITY,
+                    reporting.Groups.NETWORK,
+                    reporting.Groups.SERVICES
+            ]),
+        ] + resources)
+        return
+
+    engines_pair = _find_pair(init_block, "engines")
+    if not engines_pair:
+        # No engines no problem
+        return
+
+    engines_block = _openssl_find_block(config, engines_pair.value)
+    if not engines_block:
+        # No engines no problem
+        return
+
+    enabled_engines = []
+    # Iterate over engines directives -- they point to another block
+    for engine in engines_block.pairs:
+        name = engine.key
+        engine_block = _openssl_find_block(config, engine.value)
```

Comment:
Now, technically the name of the engine configuration block is arbitrary, too.
For example, this is a valid configuration that load the pkcs11 engine:

```conf
openssl_conf = openssl_init

[openssl_init]
engines = engines_sect

[engines_sect]
cxpf11 = cxpf11_sect

[cxpf11_sect]
engine_id = pkcs11
dynamic_path = /usr/lib64/engines-3/libpkcs11.so
MODULE_PATH = opensc-pkcs11.so
init = 1
```

I'm not sure whether we actually want to go all the way and detect this, too. It feels like over-complicating things.

Reply:
My reading of the manual page was that this is just an "alternative name".  

>  engine_id
> This is used to specify an alternate name, overriding the default name specified in the list of engines. If present, it must be first.  For example:

You are right it overrides the original id, but the search for the pkcs11 is mostly additional check -- we really care if there are any engines or not and if they are we are going to say the user to fix their config.

I would probably like to avoid going to the engine_id, but it should not be that hard to add.